Add providers API

TanoRojas opened the pull request:
@FrontEndCoffee Hey! I am creating this PR with the requested changes in https://github.com/sandwave-io/realtimeregister-php/issues/9, but I have a question that maybe makes the PR incomplete. 

When we do filter on lists, for every domain object, the code is using a general approach of sending a param named `$search`. 
I did the same for `Providers` and `Downtimes`, but I saw that `search` is not a valid filter, and we have a list of valid params for every listing. Example: https://dm.realtimeregister.com/docs/api/providers/downtime/list

I thought in adding these filters, but it is not convenient to pass so many parameters to a method. Maybe you have planned another approach for this (or maybe I am missunderstanding something :) )

I hope the question has sense. I wait for yout comments, and thanks for letting me work on this issue!

FrontEndCoffee commented:
Hi @TanoRojas, well spotted. The search field is nearly right, it should however be set on the `$query` object as the `'q'` property.

(It's a bit hidden), but the API docs contain a page that explains how all listing endpoints have pagination, and how this works: https://dm.realtimeregister.com/docs/api/listings

![Screenshot 2020-10-03 at 14 46 47](https://user-images.githubusercontent.com/16303480/94992031-63eafb00-0587-11eb-8421-960393d40831.png)

I agree with you that the filter fields would be way to many parameters, maybe in the future we could work out a way to implement this as well, if you have any ideas please let me know 😉

For now, I would love if you can check all the list functions, and make sure the key `q` is used on the query object instead of `search`. The function parameter can stay `$search` as this is nice and clear for the user of this package.

FrontEndCoffee commented:
@TanoRojas I have created a bug issue for this, you can also just fix it for the provider list endpoint(s), and fix the others later.

It's up to you, the issue is here: https://github.com/sandwave-io/realtimeregister-php/issues/23

FrontEndCoffee commented on src/Domain/Downtime.php:
```diff
@@ -0,0 +1,58 @@
+        ?string $reason,
+        Provider $provider
+    ) {
+        $this->startDate = $startDate;
```
$this->id is never set.

TanoRojas replied:
Added: https://github.com/sandwave-io/realtimeregister-php/pull/22/commits/3a455c6b171dc6179598f060e30918f0e4bd0916

FrontEndCoffee commented on src/Domain/Provider.php:
```diff
@@ -0,0 +1,38 @@
+    {
+        return new Provider(
+            $json['name'],
+            $json['tlds'] ?? []
```
This should default to null, as this means that the TLDs are not set, instead of there not being any TLDs.

TanoRojas replied:
Done: https://github.com/sandwave-io/realtimeregister-php/pull/22/commits/05d5efc506d86688ffa8f7830986ba2951e959fd

FrontEndCoffee commented on src/Api/ProvidersApi.php:
```diff
@@ -0,0 +1,90 @@
+            $query['offset'] = $offset;
+        }
+        if (! is_null($search)) {
+            $query['search'] = $search;
```
As you we talked about, `$search` should be set on `$query['q']`

TanoRojas replied:
Done: https://github.com/sandwave-io/realtimeregister-php/pull/22/commits/9bf827a5a24ca194f3bbba68b812273faecbe6ce

FrontEndCoffee commented on tests/Domain/DomainObjectTest.php:
```diff
@@ -247,6 +249,14 @@ public function parserDataSet(): array
+            'valid provider' => [
+                Provider::class,
+                include __DIR__ . '/data/provider_valid.php',
+            ],
```
Can you also add cases with only required fields, and then cases with invalid data?

TanoRojas replied:
I added a few more cases for Provider and Downtime objects: https://github.com/sandwave-io/realtimeregister-php/pull/22/commits/9743603880caf5d6044ac7abc219848654014498

FrontEndCoffee commented on tests/Domain/data/provider_valid.php:
```diff
@@ -0,0 +1,6 @@
+
+return [
+    'name' => 'Provider Name Test',
+    'tlds' => ['.com', '.net'],
```
Please change this to `['com', 'net']`

TanoRojas replied:
Done: https://github.com/sandwave-io/realtimeregister-php/pull/22/commits/9743603880caf5d6044ac7abc219848654014498#diff-d26e3d12584eb2d647e97650dde7b1a4

FrontEndCoffee reviewed (changes requested):
Thank you for this PR

Looking very good overall! There is a few small tweaks that I would like to see, after fixing those I will merge this PR 😉

TanoRojas commented:
@FrontEndCoffee Hey Jesse! thanks for the review! I fixed the mentioned errors.
I have some doubts about the extra tests cases for Downtime. In particular on how to validate correct dates provided. I just added try/catch blocks capturing everything (catching Throwable) over the Carbon constructor, and raised a TypeError exception there. 
That could be improved but maybe is good enough for the moment, but let me know! 

I will handle the issue of changing the query parameters (https://github.com/sandwave-io/realtimeregister-php/issues/23), and once this PR is approved I will take the one related to brands Api (https://github.com/sandwave-io/realtimeregister-php/issues/10) if that sounds good to you.

TanoRojas commented:
@FrontEndCoffee I think that there is actually no real need for doing the small change of `search` for `q` on another pull request, then I just pushed the change here: https://github.com/sandwave-io/realtimeregister-php/pull/22/commits/3445ba5afb4373bb4ea36c0db7fe8c221376b3dc
I understand that is better to have 1 issue => 1 PR, but maybe is good for this case. I can remove the commit and doing on another PR if you prefer that way.

FrontEndCoffee commented:
Sure thing, you can combine it in here no problem.

FrontEndCoffee reviewed (approved):
Looks great! 😄

FrontEndCoffee reviewed (changes requested):
TODO:

- [ ] Fix the `search` -> `q` issue that we talked about.

TanoRojas commented:
>     * [ ]  Fix the `search` -> `q` issue that we talked about.

Hey Jesse! Is missing something for this replacement? I replaced the `'search'` with `'q'` in all the places I found in the code here: https://github.com/sandwave-io/realtimeregister-php/commit/3445ba5afb4373bb4ea36c0db7fe8c221376b3dc

FrontEndCoffee commented:
Hi, my bad, I had the idea that you only changed it for the providers API. I see that you changed all the places :)
